### ISSUE TITLE
Distributed training with different log_vars keys among GPUs hangs the entire work process

### ISSUE THREAD

**Issue by fingertap** [open-mmlab/mmsegmentation#1034]
**Checklist**

- [x] I have searched related issues but cannot get the expected help.
- [x] The bug has not been fixed in the latest version.

**Describe the bug**

I met this bug in issue #1030. This bug is triggered when I used distributed training and some GPUs have different `log_vars` from others (e.g., containing an accuracy term only when some conditions are met).

The following is pasted from #1030:

> In the `_parse_log` function of the `mmseg.segmentors.base.BaseSegmentor`, it attempts to synchronize the loss values among all GPUs. What happened is that in this loop ([line 194](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/base.py#L194)):
>
> ```python
> for loss_name, loss_value in log_vars.items():
>    # reduce loss when distributed training
>    if dist.is_available() and dist.is_initialized():
>        loss_value = loss_value.data.clone()
>        dist.all_reduce(loss_value.div_(dist.get_world_size()))
>    log_vars[loss_name] = loss_value.item()
>```
>
>One GPU A does not have a `"roi_acc"` as `loss_name` (suppose it is the last key in `log_vars`). Then this GPU A thinks it has done all its work, and jump out of the loop. Other GPUs with the last `"roi_acc"` will try to call `torch.distributed.all_reduce`, which infinitely waits for the reply from GPU A.

**Reproduction**

1. What command or script did you run?

The complete code is very complicated. I think the following code can reproduce this issue.

```python
def forward_train(self, img, img_metas, **kwargs):
    losses = dict()
    rank = dist.get_rank()
    if rank != 6:
        losses['acc'] = torch.tensor(1.)
    return losses
```

This case is common, as the metrics may depend on the input data (e.g. only output the acc when there is an target in this image). This case is especially frequent when you have a batchsize=1. Passing 0 to the metric is not a good idea as it cannot reflect the real performance of the model.

2. Did you make any modifications on the code or config? Did you understand what you have modified?

No modification. I subclassing from `mmseg` and `mmdet`. Yes, I read the code.

3. What dataset did you use?

A custom dataset. This bug is dataset-agnostic.

**Environment**

1. Please run `python mmseg/utils/collect_env.py` to collect necessary environment information and paste it here.

This bug is environment-agnostic.

2. You may add addition that may be helpful for locating the problem, such as
    - How you installed PyTorch [e.g., pip, conda, source]
    - Other environment variables that may be related (such as `$PATH`, `$LD_LIBRARY_PATH`, `$PYTHONPATH`, etc.)

Not relevant.

**Error traceback**

It just hangs without errors. Maybe after a long time, there will be a timeout.

**Bug fix**

This is about allowing users to pass a `None` log_vars. To do this, users must guarantee that the keys in log_vars is the same for all GPUs. For those log_vars they wanted to skip, the `mmcv.runner.log_buffer.LogBuffer` should support `None` values. I would like to create a PR for this when time allows.

**Comment by MengzhangLI** [comment]
Hi, thanks for your report. 

If your time available, could you make a pr so we could merge it into master branch ASAP.

Thanks for your great contribution!

Best,

**Comment by fingertap** [comment]
As this fix may be involving modifications in `mmcv` too, I'm not sure how to connect PRs between two repos.

**Comment by MengzhangLI** [comment]
Hi, @fingertap 
Thank you sooo much for your great issue & pr.
I will close this issue and let’s keep in touch in your pr.
Best,

**Comment by TAICHIKF** [comment]
> 嗨，@fingertap 非常 感谢你的好问题和公关。 我将关闭这个问题，让我们在你的公关中保持联系。 最好的，

请问这个问题解决了吗

**Comment by MengzhangLI** [comment]
> 

Yes, related PR has fixed it. Please update latest mmseg and mmcv for better usage experience!

Best,

**Comment by MengzhangLI** [comment]
> > 嗨，@fingertap 非常 感谢你的好问题和公关。 我将关闭这个问题，让我们在你的公关中保持联系。 最好的，
> 
> 请问这个问题解决了吗

Yes, related PR has fixed it. Please update latest mmseg and mmcv for better usage experience!

Best,
